shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Val.hs

```diff
@@ -171,3 +136,120 @@ scaledMinDeposit v (Coin mv)
     -- size of the UTxO entry (ie the space the scaled minUTxOValue deposit pays)
     utxoEntrySizeWithoutVal :: Integer
     utxoEntrySizeWithoutVal = inputSize + outputSizeWithoutVal
+
+addrHashLen :: Integer
+addrHashLen = 28
+
+uint :: Integer
+uint = 5
+
+assetIdLen :: Integer
+assetIdLen = 30 -- I have no idea what this is supposed to be
+
+-- ================================================================================================
+-- There are 2 basic instances of the Val class, and most other instances are built from them
+-- from some kind of automatic inheritance from these two types.
+
+instance Val Integer where
+  zero = 0
+  x <+> y = x + y
+  n <×> x = fromIntegral n * x
+  coin x = Coin x
+  inject (Coin x) = x
+  size _ = 1
+  isZero x = x == 0
+
+-- ==================================================================
+-- The Coin instance inherits from Integer by newtype deriving
+
+deriving instance Val Coin
+
+-- ======================================================================\
+-- We can nest Map.Map over a Val t, and that inherits from t
+-- This means instances like (Map k1 (Map k2 (Map k3 t))) come for free if Val t
+
+instance (Typeable k, Show k, Ord k, Val t) => Val (Map.Map k t) where
```

Comment:
I deliberately removed the `Val` instance for `Map` because it doesn't make sense; please don't re-introduce it. 

For a start, it's not even lawful according to your own property specifications, since `coin . inject` is not an identity. It only makes sense in the context of knowing that it will be used within another instance, and worse, relies on the specific nature of that instance. It also means we split the definition of what it means to be a multi-asset over two places, and end up defining MA-specific stuff at this top level.

Reply:
In light of the light of the discussion above. I agree that it only make sense in the version of Val augmented with the coin, inject, modify operations, where we think of the Map case having a default Coin of zero. In that case the law does hold. I will endeavor to clean this up, perhaps by introducing two classes:  BoxOfInt (or some better name) and Val  where BoxOfInt is a superclass of Val